1.22 feature blog for memory qos support with cgroups v2 (#29015)
kubernetes/website: content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

---
layout: blog

> title: 'Support memory qos with cgroups v2'

Reviewer: Maybe something like:
```suggestion
title: "Alpha in Kubernetes v1.22: quality-of-service for memory resources"
```
and a slug to match?

Reviewer 2: Done

date: 2021-08-11
slug: memory-qos-cgroups-v2
---

**Authors:** Tim Xu (Tencent Cloud)

> Memory QoS is used to memory protection and usage throttle on pod / container to increase memory isolation and service quality.

Reviewer 3: ```suggestion
Memory QoS provides memory protection and usage throttling on pods and containers to increase memory isolation and service quality.
```

Reviewer: ```suggestion
Kubernetes v1.22 adds a new (alpha) feature that improves how Linux nodes
implement memory resource requests and limits. The new feature relies on
cgroups v2; most current operating system releases for Linux already provide that.
With this experimental feature,
[quality-of-service for pods and containers](/docs/tasks/configure-pod-container/quality-service-pod/)
extends to cover not just CPU time but RAM as well.
```

(I think it's useful to link to that page even though it has been overlooked for an update about the new Memory QoS feature: we can fix the task page post-release, and all the time it's still the best place for the announcement to link to).

> In traditional cgroups v1 in Kubernetes, cpu resources can be very well protected. Kubernetes uses `cpu_shares / cpu_set / cpu_quota / cpu_period` to isolate cpu. But the memory is not well guaranteed, especially at cgroup level. Kubernetes only uses `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used neither by cgroups v1 nor cgroups v2 to protect memory requests. About memory protection, we use `oom_scores` to determine order of killing container process when OOM occurs.

Reviewer 3: ```suggestion
In traditional cgroups v1 in Kubernetes, cpu resources can be very well protected. Kubernetes uses `cpu_shares`, `cpu_set`, `cpu_quota`, and `cpu_period` to isolate cpu. But the memory is not well guaranteed, especially at the cgroup level. Kubernetes only uses `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used either by cgroups v1 or cgroups v2 to protect memory requests. For memory protection, we use `oom_scores` to determine the order of killing container processes when OOM occurs.
```

Reviewer: ```suggestion
Kubernetes already has good protection for CPU resources. On Linux nodes,
Kubernetes typically sets `cpu_shares`, `cpu_set`, `cpu_quota`, and `cpu_period` to isolate
CPU and ensure fairness between workloads. For all resources, Kubernetes
allows a workload to define both a _request_ and a _limit_. Existing Kubernetes
code already instructs the container runtime to take both settings into account
for CPU resources.

Before this new feature, Kubernetes didn't do as good a job isolating workloads
from each other in terms of RAM use. The existing mechanism only used simple
calculations (`memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])`
with cgroups v1, or
`memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2)
to limit use of the `memory` resource.
The kubelet and container runtime didn't consider memory requests at all, only limits.
```

> It may cause:

Reviewer 3: ```suggestion
Killing container processes may cause:
```

> - Pod / Container memory requests can't be fully reserved, page cache is at risk of being recycled

Reviewer 3: ```suggestion
- Pod or Container memory requests to not be fully reserved and page cache to become at risk of being recycled
```

Reviewer: I'd write:
```suggestion
- The kubelet may not be able to guarantee memory resources even when you
  specify these with a `limit` or a `request` for a Pod or for one of its containers.
  Because the memory could be overcommitted at the host operating system level,
  memory pages  that are backed by non-volatile storage (such as the executable
  code for your app) could be dropped from RAM due to overall memory pressure,
   even when your Pod is running with Guaranteed QoS.
```
and I think that's also accurate technically.

Reviewer 2: OK, Have updated.

> - Pod / Container memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out

Reviewer 3: ```suggestion
- Pod or Container memory allocation to not be well protected and allocation latency frequently may occur when node memory nearly runs out
```

Reviewer: I'm not sure I understand what this means so I can't work out a good way to paraphrase it. What form of allocation is being discussed here? `sbrk()` and friends?

> - Memory overcommit of container is not throttled, there may increase risk of node memory pressure

Reviewer 3: ```suggestion
- Memory overcommit of container to not be throttled and there may increase a risk of node memory pressure
```

Reviewer: I'm not sure I understand what this means so I can't work out a good way to paraphrase it. What form of throttling is being discussed here?

> Kubernetes 1.22 brings Memory QoS to alpha to pay attention on the quality of service of memory resources. It uses cgroup v2 new capabilities `memory.min / memory.high` in memory controller to retain memory requests and throttle overcommit limits. This feature can improve memory availability and reduce memory requests reclamation, which is of great significance to ensure the stability of both workloads and nodes.

Reviewer 3: ```suggestion
Kubernetes 1.22 brings Memory QoS to alpha to pay attention to the quality of service of memory resources. It uses cgroup v2 new capabilities `memory.min` and `memory.high` in memory controller to retain memory requests and throttle overcommit limits. This feature can improve memory availability and reduce memory requests reclamation, which is of great significance to ensure the stability of both workloads and nodes.
```

Reviewer: Is this correct:
```suggestion
Kubernetes 1.22 introduces a new alpha feature for defining quality-of-service for memory
resources. This new feature relies on new cgroup v2 capabilities: `memory.min` and
`memory.high`. The kubelet's memory management subsystem uses those capabilities to
define a guaranteed amount of RAM for the cgroup overall, as well as to restrict how much
memory that control group can try to overcommit. Using memory QoS on your nodes can
help you to improve application availability. The `memory.min` setting also helps avoid
paging related to the Pods and containers that it applies to, which is important for
latency-sensitive workloads.
```
?

Reviewer 2: It's basically right. I made some minor adjustments.

## How does it work?

> It uses two interfaces `memory.min / memory.high` of cgroup v2 memory controller to work.

Reviewer 3: ```suggestion
It uses two interfaces, `memory.min` and `memory.high` of cgroup v2 memory controller, to work.
```

| File | Description |
| -------- | -------- |
| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. **We map it to `requests.memory`.** |

> | memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. **We use a formula to calculate `memory.high` depending on `limits.memory/node allocatable memory` and a memory throttling factor.** |

Reviewer 3: ```suggestion
| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. **We use a formula to calculate `memory.high`, depending on `limits.memory/node allocatable memory` and a memory throttling factor.** |
```

> | memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. **We use a formula to calculate `memory.high` depending on `limits.memory/node allocatable memory` and a memory throttling factor.** |

Reviewer 2: Have updated.

> Both container and pod are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container / pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation. 

Reviewer 3: ```suggestion
Both containers and pods are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container and pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation. 
```

> ![](https://i.imgur.com/FpR0cUK.png)

Reviewer 2: Have updated.

The formula is as follows:
```
// Container
/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
/cgroup2/kubepods/pod<UID>/<container-id>/memory.high=(pod.spec.containers[i].resources.limits[memory]/node allocatable memory)*memory throttling factor // Burstable
// Pod
/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers[i].resources.requests[memory])
// QoS ancestor cgroup

> /cgroup2/kubepods/burstable/memory.min=sum(pod[i].spec.containers[j].resources.requests[memory])

Reviewer 2: ok, will update.

```

> Cgroup v2 memory controller would retain memory requests or throttle memory usages in the pod or container cgroup based on the set value. Because effective min boundary is limited by `memory.min` values of all ancestor cgroups, kubelet also sets qos or root cgroup if necessary.

Reviewer: I'm afraid I'm not sure what point is being made here.
Is one of these true:
```suggestion
The cgroup v2 memory controller enforces memory requests and ceilings
per-pod or per-container. If you set. The design of cgroup v2 also implies
for the `memory.min` behavior to be as described, all ancestor cgroups
must have appropriate `memory.min` values. To ensure this, the kubelet sets
`memory.min` on the parent cgroup (representing the Pod).
```
```suggestion
The cgroup v2 memory controller enforces memory requests and ceilings
per-pod or per-container. If you set. The design of cgroup v2 also implies
for the `memory.min` behavior to be as described, all ancestor cgroups
must have appropriate `memory.min` values. To ensure this, the kubelet sets
`memory.min` on all ancestor cgroups up to the root.
You are responsible for ensuring that your init system (or other operating system
level tooling) does not also try to manage `memory.min` on those ancestor cgroups.
```
?

Also, should we mention `memory_recursiveprot` as a mount option?

Reviewer 2: I think the second is betther.

## How do I use it?

> All prerequisites must be met to enable the memory qos feature.

Reviewer: ```suggestion
Here are the prerequisites for enabling memory QoS on your Linux node:
```

1. Kubernetes since v1.22
2. runc since v1.0.0-rc93, containerd since 1.14
3. Linux kernel minimum version: 4.15, recommended version: 5.2+

> 4. Kernel enables cgroups v2 unified hierarchy, add `systemd.unified_cgroup_hierarchy=1` to the kernel cmdline

Reviewer: ```suggestion
4. You have enabled the cgroup v2 unified hierarchy (for example: add
   `systemd.unified_cgroup_hierarchy=1` to the kernel command line).
```

> After that, user can enable memory qos in feature gate, add `--feature-gates="...,MemoryQoS=true"` to kubelet flag.

Reviewer 3: ```suggestion
After that, the user can enable the memory qos feature gate by adding `--feature-gates="...,MemoryQoS=true"` to the kubelet flag.
```

Reviewer: ```suggestion
With those prerequisites met, you can enable the memory QoS feature gate (see
[Set kubelet parameters via a config file](/docs/tasks/administer-cluster/kubelet-config-file/).
```

## How can I learn more?

You can find more details as follows:
- KEPs
    - Memory QoS https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos
    - Cgroup v2 https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2
- PRs
    - Feature: add unified on CRI to support cgroup v2 https://github.com/kubernetes/kubernetes/pull/102578

>     - Feature: Support memory qos with cgroups v2 https://github.com/kubernetes/kubernetes/pull/102970

Reviewer 3: Is there a reason for not using markdown for the four links under KEPs and PRs? For example:

` [Memory Qos](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos)`

Reviewer: If there's documentation for this feature, I'd like to link to it.

(If we haven't documented the new behavior, why is that?)

>     - Memory QoS https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos
>     - Cgroup v2 https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2
> - PRs
>     - Feature: add unified on CRI to support cgroup v2 https://github.com/kubernetes/kubernetes/pull/102578
>     - Feature: Support memory qos with cgroups v2 https://github.com/kubernetes/kubernetes/pull/102970

Reviewer: ```suggestion
     - [Support Memory QoS with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos/#readme)
     - [cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2/#readme)
```
We usually _don't_ link directly from blogs about new features to PRs that are part of their implementation.
